File: Cartman_cipher/cartman.pas
(**      Cartman.PAS by Alexander Myasnikov                               **)
(**      Cartman Tiny block cipher for Delphi programmers                 **)
(**      Cartman encrypts 128 bit block with 512 bit key                  **)
(**      WEB:       www.darksoftware.narod.ru                             **)


Unit cartman;

{$Q-}
{$R-}


Interface

Type t128buf = array [0..3] Of longword;

Type p128buf = ^t128buf;

Procedure cartman_crypt(v: p128buf);
// Encrypt four 32 bit longwords

Procedure cartman_decrypt(v: p128buf);
// Decrypt four 32 bit longwords

Procedure cartman_setkey(k: pointer);
// Set 512 bit key (Pointer to 64 bytes or 16 longwords block)

Procedure cartman_killkey();
// Clear internal key (after encryption or decryption, just security hint)

Var key: array [0..15] Of longword;


Implementation


Function ror32(N, R:longword): longword;
asm
MOV EAX,N
MOV ECX, R
ROR EAX, CL
End;



Function rol32(N, R:longword): longword;
asm
MOV EAX,N
MOV ECX, R
ROL EAX,CL
End;






Procedure cartman_crypt(v: p128buf);

Var ki, k1, k2, k3, k4, mod1, mod2: longword;
  i: longint;


Begin

  For i:=0 To 127 Do
    Begin

      ki := i Mod 8;
      k1 := key [ki*2];
      k2 := key [ki*2+1];
      k3 := key [15-ki*2];
      k4 := key [15-ki*2+1];
      mod1 := k1 Mod 32;
      mod2 := k2 Mod 32;

      v[0]:= v[0] xor  (v[1]+v[2]) mod k1;
      v[1]:= v[1] xor  (v[2]+v[3]) mod k2;
      v[2]:= v[2] xor  (v[3]+v[0]) mod k1;
      v[3]:= v[3] xor  (v[0]+v[2]) mod k2;

      inc (v[0],  ROR32 (v[1] + k1, mod2));
      dec (v[0],  ROL32 (v[2] + k2, mod1));
      inc (v[0],  ROR32 (v[3] + k1, mod2));
      dec (v[1],  ROL32 (v[0] + k2, mod1));
      inc (v[1],  ROR32 (v[2] + k1, mod2));
      dec (v[1],  ROL32 (v[3] + k2, mod1));
      inc (v[2],  ROR32 (v[0] + k1, mod2));
      dec (v[2],  ROL32 (v[1] + k2, mod1));
      inc (v[2],  ROR32 (v[3] + k1, mod2));
      dec (v[3],  ROL32 (v[0] + k2, mod1));
      inc (v[3],  ROR32 (v[1] + k1, mod2));
      dec (v[3],  ROL32 (v[2] + k2, mod1));

      dec (v[0], (v[1] shl 6) xor (v[1] shr 9 ) + (v[1] xor k3 ) );
      inc (v[0], (v[2] shl 6) xor (v[2] shr 9 ) + (v[2] xor k4 ) );
      dec (v[0], (v[3] shl 6) xor (v[3] shr 9 ) + (v[3] xor k3 ) );
      inc (v[1], (v[0] shl 6) xor (v[0] shr 9 ) + (v[0] xor k4 ) );
      dec (v[1], (v[2] shl 6) xor (v[2] shr 9 ) + (v[2] xor k3 ) );
      inc (v[1], (v[3] shl 6) xor (v[3] shr 9 ) + (v[3] xor k4 ) );
      dec (v[2], (v[0] shl 6) xor (v[0] shr 9 ) + (v[0] xor k3 ) );
      inc (v[2], (v[1] shl 6) xor (v[1] shr 9 ) + (v[1] xor k4 ) );
      dec (v[2], (v[3] shl 6) xor (v[3] shr 9 ) + (v[3] xor k3 ) );
      inc (v[3], (v[0] shl 6) xor (v[0] shr 9 ) + (v[0] xor k4 ) );
      dec (v[3], (v[1] shl 6) xor (v[1] shr 9 ) + (v[1] xor k3 ) );
      inc (v[3], (v[2] shl 6) xor (v[2] shr 9 ) + (v[2] xor k4 ) );
    End;


End;


Procedure cartman_decrypt(v: p128buf);

Var ki, k1, k2, k3, k4, mod1, mod2: longword;
  i: longint;


Begin

  For i:=127 Downto 0 Do
    Begin

      ki := i Mod 8;
      k1 := key [ki*2];
      k2 := key [ki*2+1];
      k3 := key [15-ki*2];
      k4 := key [15-ki*2+1];
      mod1 := k1 Mod 32;
      mod2 := k2 Mod 32;


      dec (v[3], (v[2] shl 6) xor (v[2] shr 9 ) + (v[2] xor k4 ) );
      inc (v[3], (v[1] shl 6) xor (v[1] shr 9 ) + (v[1] xor k3 ) );
      dec (v[3], (v[0] shl 6) xor (v[0] shr 9 ) + (v[0] xor k4 ) );
      inc (v[2], (v[3] shl 6) xor (v[3] shr 9 ) + (v[3] xor k3 ) );
      dec (v[2], (v[1] shl 6) xor (v[1] shr 9 ) + (v[1] xor k4 ) );
      inc (v[2], (v[0] shl 6) xor (v[0] shr 9 ) + (v[0] xor k3 ) );
      dec (v[1], (v[3] shl 6) xor (v[3] shr 9 ) + (v[3] xor k4 ) );
      inc (v[1], (v[2] shl 6) xor (v[2] shr 9 ) + (v[2] xor k3 ) );
      dec (v[1], (v[0] shl 6) xor (v[0] shr 9 ) + (v[0] xor k4 ) );
      inc (v[0], (v[3] shl 6) xor (v[3] shr 9 ) + (v[3] xor k3 ) );
      dec (v[0], (v[2] shl 6) xor (v[2] shr 9 ) + (v[2] xor k4 ) );
      inc (v[0], (v[1] shl 6) xor (v[1] shr 9 ) + (v[1] xor k3 ) );

      inc (v[3],  ROL32 (v[2] + k2, mod1));
      dec (v[3],  ROR32 (v[1] + k1, mod2));
      inc (v[3],  ROL32 (v[0] + k2, mod1));
      dec (v[2],  ROR32 (v[3] + k1, mod2));
      inc (v[2],  ROL32 (v[1] + k2, mod1));
      dec (v[2],  ROR32 (v[0] + k1, mod2));
      inc (v[1],  ROL32 (v[3] + k2, mod1));
      dec (v[1],  ROR32 (v[2] + k1, mod2));
      inc (v[1],  ROL32 (v[0] + k2, mod1));
      dec (v[0],  ROR32 (v[3] + k1, mod2));
      inc (v[0],  ROL32 (v[2] + k2, mod1));
      dec (v[0],  ROR32 (v[1] + k1, mod2));

      v[3]:= v[3] xor  (v[0]+v[2]) mod k2;
      v[2]:= v[2] xor  (v[3]+v[0]) mod k1;
      v[1]:= v[1] xor  (v[2]+v[3]) mod k2;
      v[0]:= v[0] xor  (v[1]+v[2]) mod k1;

    End;
End;




Procedure cartman_setkey(k: pointer);
Begin
  move(k^,key,64);
End;


Procedure cartman_killkey();
Begin
  fillchar(key, 64, 0);
  fillchar(key, 64, $FF);
End;


End.

File: Cartman_cipher/libcartman/cartman.c
/*      Cartman.C by Alexander Myasnikov                               **/
/*      Cartman Tiny block cipher for Delphi programmers                 **/
/*      Cartman encrypts 128 bit block with 512 bit key                  **/
/*      WEB:       www.darksoftware.narod.ru                             **/


#include <string.h>
#include <stdlib.h>


#define u32 unsigned long

#define ROR32(x, n)  _lrotr(x,n)
#define ROL32(x, n)  _lrotl(x,n)

u32 key[16]= {0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0};


void __stdcall __export crypt (u32 *v)
{
u32 ki, k1, k2, k3, k4, mod1, mod2;
long i;


  for (i=0; i<=127;i++)
{

      ki = i%8;
      k1 = key [ki*2];
      k2 = key [ki*2+1];
      k3 = key [15-ki*2];
      k4 = key [15-ki*2+1];
      mod1 = k1%32;
      mod2 = k2%32;


      v[0]^=  (v[1]+v[2]) % k1;
      v[1]^=  (v[2]+v[3]) % k2;
      v[2]^=  (v[3]+v[0]) % k1;
      v[3]^=  (v[0]+v[2]) % k2;



      v[0]+=  ROR32 (v[1] + k1, mod2);
      v[0]-=  ROL32 (v[2] + k2, mod1);
      v[0]+=  ROR32 (v[3] + k1, mod2);
      v[1]-=  ROL32 (v[0] + k2, mod1);
      v[1]+=  ROR32 (v[2] + k1, mod2);
      v[1]-=  ROL32 (v[3] + k2, mod1);
      v[2]+=  ROR32 (v[0] + k1, mod2);
      v[2]-=  ROL32 (v[1] + k2, mod1);
      v[2]+=  ROR32 (v[3] + k1, mod2);
      v[3]-=  ROL32 (v[0] + k2, mod1);
      v[3]+=  ROR32 (v[1] + k1, mod2);
      v[3]-=  ROL32 (v[2] + k2, mod1);

      v[0]+= ((v[1] << 6) ^ (v[1] >> 9 ) + (v[1] ^ k4 ));
      v[0]-= ((v[2] << 6) ^ (v[2] >> 9 ) + (v[2] ^ k3 ));
      v[0]+= ((v[3] << 6) ^ (v[3] >> 9 ) + (v[3] ^ k4 ));
      v[1]-= ((v[0] << 6) ^ (v[0] >> 9 ) + (v[0] ^ k3 ));
      v[1]+= ((v[2] << 6) ^ (v[2] >> 9 ) + (v[2] ^ k4 ));
      v[1]-= ((v[3] << 6) ^ (v[3] >> 9 ) + (v[3] ^ k3 ));
      v[2]+= ((v[0] << 6) ^ (v[0] >> 9 ) + (v[0] ^ k4 ));
      v[2]-= ((v[1] << 6) ^ (v[1] >> 9 ) + (v[1] ^ k3 ));
      v[2]+= ((v[3] << 6) ^ (v[3] >> 9 ) + (v[3] ^ k4 ));
      v[3]-= ((v[0] << 6) ^ (v[0] >> 9 ) + (v[0] ^ k3 ));
      v[3]+= ((v[1] << 6) ^ (v[1] >> 9 ) + (v[1] ^ k4 ));
      v[3]-= ((v[2] << 6) ^ (v[2] >> 9 ) + (v[2] ^ k3 ));

  

}

 
}


void __stdcall __export  decrypt (u32 *v)
{

u32  ki, k1, k2, k3, k4, mod1, mod2;
long i;


  for (i=127; i>=0;i--)
{

      ki = i%8;
      k1 = key [ki*2];
      k2 = key [ki*2+1];
      k3 = key [15-ki*2];
      k4 = key [15-ki*2+1];
      mod1 = k1%32;
      mod2 = k2%32;


//////

      v[3]+= ((v[2] << 6) ^ (v[2] >> 9 ) + (v[2] ^ k3 ));
      v[3]-= ((v[1] << 6) ^ (v[1] >> 9 ) + (v[1] ^ k4 ));
      v[3]+= ((v[0] << 6) ^ (v[0] >> 9 ) + (v[0] ^ k3 ));
      v[2]-= ((v[3] << 6) ^ (v[3] >> 9 ) + (v[3] ^ k4 ));
      v[2]+= ((v[1] << 6) ^ (v[1] >> 9 ) + (v[1] ^ k3 ));
      v[2]-= ((v[0] << 6) ^ (v[0] >> 9 ) + (v[0] ^ k4 ));
      v[1]+= ((v[3] << 6) ^ (v[3] >> 9 ) + (v[3] ^ k3 ));
      v[1]-= ((v[2] << 6) ^ (v[2] >> 9 ) + (v[2] ^ k4 )) ;
      v[1]+= ((v[0] << 6) ^ (v[0] >> 9 ) + (v[0] ^ k3 )) ;
      v[0]-= ((v[3] << 6) ^ (v[3] >> 9 ) + (v[3] ^ k4 )) ;
      v[0]+= ((v[2] << 6) ^ (v[2] >> 9 ) + (v[2] ^ k3 )) ;  
      v[0]-= ((v[1] << 6) ^ (v[1] >> 9 ) + (v[1] ^ k4 )) ;


      v[3]+=  ROL32 (v[2] + k2, mod1);
      v[3]-=  ROR32 (v[1] + k1, mod2);
      v[3]+=  ROL32 (v[0] + k2, mod1);
      v[2]-=  ROR32 (v[3] + k1, mod2);
      v[2]+=  ROL32 (v[1] + k2, mod1);
      v[2]-=  ROR32 (v[0] + k1, mod2);
      v[1]+=  ROL32 (v[3] + k2, mod1);
      v[1]-=  ROR32 (v[2] + k1, mod2);
      v[1]+=  ROL32 (v[0] + k2, mod1);
      v[0]-=  ROR32 (v[3] + k1, mod2);
      v[0]+=  ROL32 (v[2] + k2, mod1);
      v[0]-=  ROR32 (v[1] + k1, mod2);

      v[3]^=  (v[0]+v[2]) % k2;
      v[2]^=  (v[3]+v[0]) % k1;
      v[1]^=  (v[2]+v[3]) % k2;
      v[0]^=  (v[1]+v[2]) % k1;


    
}


}




void __stdcall __export setup (u32 *data)
{
 memmove(&key[0],&data[0], 64);
}


void __stdcall __export cartman_killkey()
{
  memset(key, 0, 64);
  memset(key, 0, 0xFF);
}



